template/template_test.go

```diff
@@ -172,6 +172,71 @@ func Test_ExecuteComplexMessage(t *testing.T) {
 	}
 }
 
+func Test_Matcher(t *testing.T) {
+	t.Parallel()
+
+	tests := []struct {
+		name      string
+		inputStr  string
+		inputMaps []map[string]any
+		expected  []string
+		funcs     []registry.Func
+	}{
+		{
+			name:     "matcher string",
+			inputStr: `.match { $n :string } no {{no apples}} one {{{ $n } apple}} * {{{ $n } apples}}`,
+			inputMaps: []map[string]any{
+				{"n": "no"},
+				{"n": "one"},
+				{"n": "many"},
+			},
+			expected: []string{"no apples", "one apple", "many apples"},
+		},
+		{
+			name: "Pattern Selection with string annotation",
+			//nolint:dupword
+			inputStr: ".match {$foo :string} {$bar :string} bar bar {{All bar}} foo foo {{All foo}} * * {{Otherwise}}",
+			inputMaps: []map[string]any{
+				{"foo": "foo", "bar": "bar"},
+			},
+			expected: []string{"Otherwise"},
+		},
+		{
+			name:      "Pattern Selection with Multiple Variants",
+			inputStr:  ".match {$foo :string} {$bar :string} * bar {{Any and bar}}foo * {{Foo and any}} foo bar {{Foo and bar}} * * {{Otherwise}}", //nolint:lll
+			inputMaps: []map[string]any{{"foo": "foo", "bar": "bar"}},
+			expected:  []string{"Foo and bar"},
+		},
+		{
+			name:      "Plural Format Selection",
+			inputStr:  ".match {$count :string} one {{Category match}} 1 {{Exact match}} *   {{Other match}}",
+			inputMaps: []map[string]any{{"count": "1"}},
+			expected:  []string{"Exact match"},
+		},
+	}
+
+	for _, tt := range tests {
+		if len(tt.inputMaps) != len(tt.expected) {
+			t.Error("Arguments and expected results should have the same length")
+		}
+
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+
+			template, err := New().Parse(tt.inputStr)
+			require.NoError(t, err)
+
+			for i, inputMap := range tt.inputMaps {
+				actual, err := template.Sprint(inputMap)
+				require.NoError(t, err)
+
+				require.Equal(t, tt.expected[i], actual)
```

Comment:
Nested table testing.

```suggestion
				t.Run(tt.expected[i], func (t *testing.T) {
					t.Parallel()

					actual, err := template.Sprint(inputMap)
					require.NoError(t, err)

					require.Equal(t, tt.expected[i], actual)
				})
```